### PR TITLE
Token File

### PR DESCRIPTION
**Pull request by kemphack**
This change is [Reviewable](https://reviewable.io/reviews/mcl1303/taskbot/15)

### CODE REVIEW

File: app/Main.hs [hunk 1]
```diff
@@ -1,10 +1,29 @@
...
+module Main (main) where
+
+import           Control.Exception (IOException, catch)
+import           System.Exit       (exitFailure)
+import           System.IO         (hPutStr, readFile, stderr)
+import           System.IO.Error   (isDoesNotExistError)
+
+import           TelegramApi       (sendMessage)
+
+putLog a = hPutStr stderr (a ++ "\n")
```
Comment (cblp) [inline app/Main.hs]: *hPutStrLn*

File: app/Main.hs [hunk 2]
```diff
@@ -1,10 +1,29 @@
...
+              putLog (show e)
+              exitFailure
 
 main :: IO ()
 main = do
-    token <- getLine
-    messageText <- getLine
-    mes <- sendMessage token messageText
-    print $ mes
+    str <- loadToken ""
```
Comment (cblp) [inline app/Main.hs]: Странное имя файла. Это работает?

File: app/Main.hs [hunk 3]
```diff
@@ -1,10 +1,29 @@
...
+    print $ str
```
Comment (cblp) [inline app/Main.hs]: Зачем убрал всю функциональность?

File: app/Main.hs [hunk 4]
```diff
@@ -1,10 +1,29 @@
...
+
+loadToken :: String -> IO String
+loadToken fileName = do
+    catch (readFile fileName) handler
+  where
+      handler e
+          |isDoesNotExistError e = do
+              putLog "Couldn't find token file."
+              exitFailure
+          |otherwise = do
```
Comment (cblp) [inline app/Main.hs]: Проверку ровно одного условия лучше записать через *if-then-else*.

**Comment by kemphack** [comment]
*[app/Main.hs, line 29 at r1](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-Kc4QX0onDMrySLh0Vya-r1-29:-KcFW2baBwwcH9_xNFRT:bqpt6x5) ([raw file](https://github.com/mcl1303/taskbot/blob/3a7c7a1425b96ee5ab3d7c97a93a586d6615c890/app/Main.hs#L29)):*
<details><summary><i>Previously, cblp (Yuriy Syrovetskiy) wrote…</i></summary><blockquote>

Зачем убрал всю функциональность?
</blockquote></details>

Этот пуш я сделал еще давно, я ветвился не от смерженного эхобота.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mcl1303/taskbot/15)*

**Comment by kemphack** [comment]
*[app/Main.hs, line 12 at r1](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-Kc4QX0onDMrySLh0Vya-r1-12:-KcFYZ4ch5mGgj0VSmug:b-uyl6ub) ([raw file](https://github.com/mcl1303/taskbot/blob/3a7c7a1425b96ee5ab3d7c97a93a586d6615c890/app/Main.hs#L12)):*
<details><summary><i>Previously, cblp (Yuriy Syrovetskiy) wrote…</i></summary><blockquote>

*hPutStrLn*
</blockquote></details>

По стандарту такая функция не определена. (есть только putStrLn)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mcl1303/taskbot/15)*

**Comment by cblp** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[app/Main.hs, line 12 at r1](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-Kc4QX0onDMrySLh0Vya-r1-12:-KcHSqljegYgyEssWzby:b-i24vin) ([raw file](https://github.com/mcl1303/taskbot/blob/3a7c7a1425b96ee5ab3d7c97a93a586d6615c890/app/Main.hs#L12)):*
<details><summary><i>Previously, kemphack wrote…</i></summary><blockquote>

По стандарту такая функция не определена. (есть только putStrLn)
</blockquote></details>

По какому стандарту? Вот же она — https://hackage.haskell.org/package/base-4.9.1.0/docs/System-IO.html#v:hPutStrLn

---

*[app/Main.hs, line 29 at r1](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-Kc4QX0onDMrySLh0Vya-r1-29:-KcHTWTL6F_A1ajP79I1:b-gd5wul) ([raw file](https://github.com/mcl1303/taskbot/blob/3a7c7a1425b96ee5ab3d7c97a93a586d6615c890/app/Main.hs#L29)):*
<details><summary><i>Previously, kemphack wrote…</i></summary><blockquote>

Этот пуш я сделал еще давно, я ветвился не от смерженного эхобота.
</blockquote></details>

Теперь исправляй

---

*[app/Main.hs, line 31 at r2](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-KcHTbIOvkm7PhGpuAFe:-KcHTbIOvkm7PhGpuAFf:b-rcv58l) ([raw file](https://github.com/mcl1303/taskbot/blob/065069fa955bba3ea965f404c4b4c0ed982fd8e8/app/Main.hs#L31)):*
> ```Haskell
>     print str
>   where
>   	filePath = "token.txt"
> ```

отступ надо делать пробелами

---

*[app/Main.hs, line 31 at r2](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-KcHTxX7Ie9pvb6m95wW:-KcHTxX7Ie9pvb6m95wX:b-hb5mkk) ([raw file](https://github.com/mcl1303/taskbot/blob/065069fa955bba3ea965f404c4b4c0ed982fd8e8/app/Main.hs#L31)):*
> ```Haskell
>     print str
>   where
>   	filePath = "token.txt"
> ```

это плохое имя переменной, оно не отражает суть

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mcl1303/taskbot/15)*

**Comment by cblp** [comment]
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[app/Main.hs, line 19 at r3](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-KcZ6ZQqOS47WwepsHSp:-KcZ6ZQqOS47WwepsHSq:b-tkirer) ([raw file](https://github.com/mcl1303/taskbot/blob/2cef4331da1b7b75f83aaaa866613666213fba43/app/Main.hs#L19)):*
> ```Haskell
>     :: String -- ^ Token
>     -> [Update] -- ^ List of updates
>     -> Maybe Int -- ^ Offset(current update id)
> ```

Пробел перед открывающейся скобкой

---

*[app/Main.hs, line 25 at r3](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-KcZ6K5Pw-LLU-IuqpIw:-KcZ6K5Pw-LLU-IuqpIx:bilgvsl) ([raw file](https://github.com/mcl1303/taskbot/blob/2cef4331da1b7b75f83aaaa866613666213fba43/app/Main.hs#L25)):*
> ```Haskell
>         []   -> case offset of
>             Nothing -> pure offset
>             Just a  -> pure (Just a)
> ```

В чём смысл этого кейса?

---

*[app/Main.hs, line 38 at r3](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-KcZ6y_PvJfk7MUbbmpw:-KcZ6y_PvJfk7MUbbmpx:bguwgtw) ([raw file](https://github.com/mcl1303/taskbot/blob/2cef4331da1b7b75f83aaaa866613666213fba43/app/Main.hs#L38)):*
> ```Haskell
>     getMessageText a = case msgText a of
>         Nothing      -> ""
>         Just message -> message
> ```

Перепиши кейс через *fromMaybe*

---

*[app/Main.hs, line 51 at r3](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-KcZ7g_ZJi_zi1AVc24j:-KcZ7g_ZJi_zi1AVc24k:bfajcx8) ([raw file](https://github.com/mcl1303/taskbot/blob/2cef4331da1b7b75f83aaaa866613666213fba43/app/Main.hs#L51)):*
> ```Haskell
>         Just updates  -> do
>             newOffset <- processUpdates token updates curOffset
>             bot token newOffset
> ```

`bot token someOffset` — общая часть, надо вынести из кейса

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mcl1303/taskbot/15)*

**Comment by cblp** [comment]
Reviewed 1 of 3 files at r3.
Review status: 1 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[app/Main.hs, line 34 at r3](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-KcZBvRrZatKY0B3qUWY:-KcZBvRrZatKY0B3qUWZ:b-4gypzs) ([raw file](https://github.com/mcl1303/taskbot/blob/2cef4331da1b7b75f83aaaa866613666213fba43/app/Main.hs#L34)):*
> ```Haskell
>                         (chtId (msgChat message))
>             writeParam updateIdFile (updUpdate_id x)
>             processUpdates token xs (Just (updUpdate_id x + 1))
> ```

Вижу, что ты освоился с рекурсией, но в данном случае явная рекурсия не лучшее решение. Перепиши через *for_*.

---

*[app/Main.hs, line 63 at r3](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-KcZBBmMDo3b9awvyi4U:-KcZBBmMDo3b9awvyi4V:b7smdh7) ([raw file](https://github.com/mcl1303/taskbot/blob/2cef4331da1b7b75f83aaaa866613666213fba43/app/Main.hs#L63)):*
> ```Haskell
>         else do
>             putLog (show e)
>             exitFailure
> ```

`exitFailure` — общая часть, надо вынести из кейса

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mcl1303/taskbot/15)*

**Comment by kemphack** [comment]
*[app/Main.hs, line 34 at r3](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-KcZBvRrZatKY0B3qUWY:-KcfcIumTrZZaO0yD707:b-b7c3dn) ([raw file](https://github.com/mcl1303/taskbot/blob/2cef4331da1b7b75f83aaaa866613666213fba43/app/Main.hs#L34)):*
<details><summary><i>Previously, cblp (Yuriy Syrovetskiy) wrote…</i></summary><blockquote>

Вижу, что ты освоился с рекурсией, но в данном случае явная рекурсия не лучшее решение. Перепиши через *for_*.
</blockquote></details>

А как я нужный оффсет буду возвращать?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mcl1303/taskbot/15)*

**Comment by cblp** [comment]
Review status: 1 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[app/Main.hs, line 34 at r3](https://reviewable.io:443/reviews/mcl1303/taskbot/15#-KcZBvRrZatKY0B3qUWY:-Kcg-ruJ1_4V8c7NJVah:b-ydlzi7) ([raw file](https://github.com/mcl1303/taskbot/blob/2cef4331da1b7b75f83aaaa866613666213fba43/app/Main.hs#L34)):*
<details><summary><i>Previously, kemphack wrote…</i></summary><blockquote>

А как я нужный оффсет буду возвращать?
</blockquote></details>

Смещение ты можешь вычислить отдельно от цикла с помощью *lastMay*.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mcl1303/taskbot/15)*

**Pull request closed without merging**